In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from utils.funciones_filtrado import tipo_dato_x_columna, nulos_x_columna, unicos_x_columna

In [18]:
base_ventas = pd.read_excel('ventas.xlsx')
print(base_ventas.head())

no_registros, no_columnas = base_ventas.shape
print("\n")
print(f"Número de productos registrados: {no_registros}, número de columnas de la tabla: {no_columnas}")

  venta_id       fecha       sucursal          producto categoria  cantidad  \
0    V0228  2024-01-24  Puebla  Norte           Brownie  Alimento         1   
1    V0064  10/08/2024    ANGELOPOLIS  Muffin Chocolate  Alimento         2   
2    V0128  03/15/2024   puebla norte             Latte      Café         2   
3    V0137  09/17/2024  Puebla Centro          Espresso      Café         4   
4    V0023  08/20/2024     San Andrés  Muffin Chocolate  Alimento         1   

  precio_unitario   empleado    metodo_pago  
0              28     Ana P.       efectivo  
1              55   Sofia T.        Tarjeta  
2              32   Diana F.        Tarjeta  
3              65     Ana P.        TARJETA  
4              18  Carlos R.  transferencia  


Número de productos registrados: 445, número de columnas de la tabla: 9


Notamos rápido en una inspección que la columna 'fecha' al menos el primer registro está escrito de manera diferente a las demás, situación que se tiene que corregir debido a que al tratar de hacer una seríe de tiempo puede ser uno de los motivos por el que se falle.

En la columna sucursal, notamos que la columna 'sucursal', hay una tienda que es la misma pero se escribió incorrectamente, 'Puebla Norte' y 'puebla norte', de igual manera tenemos la columna 'metodo_pago' que es el mismo método pero diferente forma de escribirlo.

In [5]:
print(tipo_dato_x_columna(base_ventas))

           columna    tipo
0         venta_id  object
1            fecha  object
2         sucursal  object
3         producto  object
4        categoria  object
5         cantidad   int64
6  precio_unitario  object
7         empleado  object
8      metodo_pago  object


Observamos que la columna 'fecha' no esta con su tipo correspondiente y podemos atribuir un poco debido a que existen formas diferentes en como se escribió la fecha.
La columna 'precio_unitario' esa esperaríamos que estuviera en tipo numérico pero vemos que está con otro tipo, posiblemente exista algún caractar no numerico en un registro que este afectando a toda la columna.

In [6]:
print('Número de valores nulos por cada columna')
print(nulos_x_columna(base_ventas))


print('\n')

print('Número de valores únicos por cada columna')
print(unicos_x_columna(base_ventas))

Número de valores nulos por cada columna
           columna  cantidad de valores nulos
0         venta_id                          0
1            fecha                          5
2         sucursal                          3
3         producto                          2
4        categoria                          0
5         cantidad                          0
6  precio_unitario                          3
7         empleado                          7
8      metodo_pago                          0


Número de valores únicos por cada columna
           columna  cantidad de valores unicos
0         venta_id                         400
1            fecha                         332
2         sucursal                          21
3         producto                          15
4        categoria                           4
5         cantidad                           5
6  precio_unitario                          46
7         empleado                           8
8      metodo_pago              

En este resumen podemos notar varias cuestiones:
1. Si bien la columna 'venta_id' no tiene valores nulos, pero si id repetidos esto debido a que la cantidad de registros de ventas, no es la misma que de valores unicos de id's, ya que la tabla tiene 445 registros y hay 400 id's diferentes, posiblemente esto nos este indicar que en una misma venta se hayan vendido diferentes productos o el registro se duplico por lo que hay que averiguar si existe diferente producto para un mismo 'id'.
2. En la columna 'fecha' hay vacios es decir, no sabremos cuando se vendieron ciertos productos, además vemos que sólo hay 332 registros diferentes y esto nos hace pensar que hubo al menos en un día, más de una venta.
3. Notamos que en 'sucursal' tenemos 21 sucursales distintas, pero por la primera inspección esto es posiblemente debido a que para una misma sucursal se haya escrito de diferentes formas.
4. Veremos si en 'producto' exista alguno mla escrito, pero para 'categoria' por la ficha técnica sabemos que sólo hay 4, entonces podemos estar seguros que esa columna está bien.
5. Podemos sospechar que la columna 'cantidad', como sólo hay 5 valores, y su tipo es numérico, pues solo contenga valores del 1 al 5.
6. Vemos que para precio unitario existen valores nulos/vacios, pero lo más interesante es que si existe 15 productos diferentes, debería de haber 15 precios_unitarios diferentes, hay casi el triple de valores unitarios, es decir que para un mismo producto hay al menos 2 diferentes precios, esto quiero decir, que hay una inconsistencia, pero habria que averiguar si es por el tamaño, pero para esto en lugar que la columna producto este el nombre, sería mejor el id del producto, para que así nos diga todas las caractisticas del producto.
7. En la columna de empleados, notamos que tenemos al menos 8 empleados diferentes, pero que en 7 registros no tenemos un nombre de un empleado de quien hizo la venta.
8. Y por último notamos que en 'metodo_pago' hay 6 metodos diferentes de pago, pero posiblemente sean menos debido a que al principio notamos que habian metodos iguales pero escritos de diferente mnaera.


In [8]:
# analizamos las ventas repetidas
ventas_repetido = base_ventas[base_ventas['venta_id'].duplicated(keep=False)].sort_values(by='venta_id')
ventas_repetido

,venta_id,fecha,sucursal,producto,categoria,cantidad,precio_unitario,empleado,metodo_pago
242,V0002,2024-01-17,Pue. Centro,Latte Vainilla,Café,5,55,Pedro H.,TARJETA
254,V0002,2024-01-17,Pue. Centro,Latte Vainilla,Café,5,55,Pedro H.,TARJETA
18,V0005,02/10/2024,Cho.,Té Negro,Té,2,30,Luis M.,Tarjeta
429,V0005,02/10/2024,Cho.,Té Negro,Té,2,30,NaN,Tarjeta
182,V0014,2024-09-07,PUEBLA CENTRO,Brownie,Alimento,5,35,María G.,Tarjeta
...,...,...,...,...,...,...,...,...,...
386,V0188,20 de October de 2024,Angélopolis,Pay de Queso,Alimento,1,30,Sofia T.,Tarjeta
190,V0188,20 de October de 2024,Angélopolis,NaN,Alimento,1,30,Sofia T.,Tarjeta
440,V0188,20 de October de 2024,Angélopolis,Pay de Queso,Alimento,1,30,Sofia T.,Tarjeta
409,V0190,2024-05-04,S. Andrés,Latte Vainilla,Café,4,55,María G.,efectivo


Podemos notar que para los valores que se repiten nos hace sospechar que los demas registros en las demás columnas los valores se repetien es decir, se duplicó la venta, por lo que haremos la suposición de que si se repite un id, es porque el registro se duplicó, esto en realidad es una suposición ambigua ya que no se sabe bien las reglas del negocio y para realizar un analisis expres se tomará esta decisión.

Como posible consulta que se puede realizar para saber si efectivamente todos los id duplicados tienen en las demás columnas tienen registros similares, se me ocurre empezar con un group_by en 'id', y posteriormente que si no es un NaN, vea que si todas las demas columnas tienen valores duplicados, y si hay un registros que no tiene en más de una columna valores no duplicados, que devuelva ese id, para analizarlo.

In [9]:
fecha_venta = base_ventas['fecha']
fecha_venta.head(10)

0               2024-01-24
1               10/08/2024
2               03/15/2024
3               09/17/2024
4               08/20/2024
5               04/14/2024
6    09 de October de 2024
7               2024-08-30
8               07/03/2024
9    17 de January de 2024
Name: fecha, dtype: object

Obtenemos una muestra más grande de la columna 'fecha' con el obejtivo de saber como escriben la fecha y al parecer, hay 3 tipos en como escriben la fecha, por lo que hay que homogenizar el formato de la fecha

In [10]:
base_ventas['sucursal'].unique()

array(['Puebla  Norte', 'ANGELOPOLIS', 'puebla norte', 'Puebla Centro',
       'San Andrés', 'Cholula', 'San Andres', 'Angelópolis',
       'Puebla centro', 'Pue. Norte', 'Angélopolis', 'Pue. Centro',
       'san andrés', 'Cho.', 'cholula', 'CHOLULA', 'Angelopolis',
       'Puebla Norte', nan, 'PUEBLA CENTRO', 'S. Andrés',
       ' Puebla Centro'], dtype=object)

Como podemos observar hay muchas formas en la que se escribió una misma sucursal, por lo que habrá que solo quedarnos con solo una forma de escribir cada sucursal

In [11]:
base_ventas['producto'].unique()

array(['Brownie', 'Muffin Chocolate', 'Latte', 'Espresso', 'Croissant',
       'Muffin Arándano', 'Té Negro', 'Cappuccino', 'Americano Grande',
       'Americano', 'Frappé Chocolate', 'Pay de Queso', 'Frappé Caramelo',
       'Latte Vainilla', 'Té Verde', nan], dtype=object)

Por una inspección rápida vemos que aquí no hay diferentes formas de escribir un mismo producto, por lo que todos son verdaderamente unicos.

In [12]:
base_ventas['categoria'].unique()

array(['Alimento', 'Café', 'Té', 'Frío'], dtype=object)

Como se sospechaba al inicio, en la columna 'categoría' solo están las 4 categorías que se habian establecido. 

In [13]:
base_ventas['cantidad'].unique()

array([1, 2, 4, 3, 5], dtype=int64)

De igual forma sólo hay 5 diferentes formas en categoría y todas son númericas

In [15]:
#vemos aquellos registros que no sean númericos
no_numero = base_ventas[pd.to_numeric(base_ventas['precio_unitario'], errors='coerce').isna()]
no_numero.head()

,venta_id,fecha,sucursal,producto,categoria,cantidad,precio_unitario,empleado,metodo_pago
23,V0151,2024-12-02,CHOLULA,Muffin Chocolate,Alimento,5,$30 USD,Sofia T.,efectivo
24,V0349,2024-07-10,Angelopolis,Croissant,Alimento,1,$28 USD,Ana P.,TARJETA
30,V0285,2024-10-20,Pue. Norte,Muffin Chocolate,Alimento,1,$28 USD,Ana P.,Efectivo
36,V0054,2024-04-24,Cho.,Americano,Café,2,$25,Ana P.,efectivo
66,V0235,2024-06-30,cholula,Té Negro,Té,4,$130 USD,Pedro H.,Tarjeta


Notamos que en la columna 'precio_unitario' como ya habiamos visto, no todos eran numericos y con esto podemos distinguir que es porque tienen el signo de dinero '$' además de la moneda en la se está vendiendo (aunque considero que a lo mejor son en pesos mexicanos, ya que en el primer registro pagar por un muffin de chocolate a 30 dolares, para mi, es demasiado)

In [20]:
base_ventas['empleado'].unique()

array(['Ana P.', 'Sofia T.', 'Diana F.', 'Carlos R.', 'Pedro H.',
       'José L.', 'María G.', 'Luis M.', nan], dtype=object)

Notamos que hay consistencia en la columna de 'empleados' ya que no existen diferentes formas de escribir a un mismo empleado.

In [22]:
base_ventas['metodo_pago'].unique()

array(['efectivo', 'Tarjeta', 'TARJETA', 'transferencia', 'Efectivo',
       'Transferencia'], dtype=object)

Notamos nuevamente que en esta columna existen diferentes formas de escribir una misma forma de pago, por lo que hay estandarizar, la forma de registrar los métodos de pago.